lib/Service/YAPServer.hpp

```diff
@@ -0,0 +1,594 @@
+
+            /* DLC = 0 means that the channel does not exist. */
+            if ((0U != dlc) && (MAX_RX_ATTEMPTS >= m_rxAttempts))
+            {
+                expectedBytes = (dlc - (m_receivedBytes - HEADER_LEN));
+                m_rxAttempts++;
+            }
+        }
+
+        /* Are we expecting to read anything? */
+        if (0U != expectedBytes)
+        {
+            /* Read the required amount of bytes, if available. */
+            if (expectedBytes <= m_stream.available())
+            {
+                m_receivedBytes += m_stream.readBytes(&m_receiveFrame.raw[m_receivedBytes], expectedBytes);
+            }
+
+            /* Frame has been received. */
+            if ((0U != dlc) && ((HEADER_LEN + dlc) == m_receivedBytes))
+            {
+                if (isFrameValid(m_receiveFrame))
+                {
+                    uint8_t channelArrayIndex = (m_receiveFrame.fields.header.headerFields.m_channel - 1U);
+
+                    /* Differenciate between Control and Data Channels. */
+                    if (CONTROL_CHANNEL_NUMBER == m_receiveFrame.fields.header.headerFields.m_channel)
+                    {
+                        callbackControlChannel(m_receiveFrame.fields.payload.m_data, CONTROL_CHANNEL_PAYLOAD_LENGTH);
+                    }
+                    else if (nullptr != m_dataChannels[channelArrayIndex].m_callback)
+                    {
+                        /* Callback */
+                        m_dataChannels[channelArrayIndex].m_callback(m_receiveFrame.fields.payload.m_data, dlc);
+                    }
+                }
+
+                /* Frame received. Cleaning! */
+                clearLocalRxBuffers();
+            }
+        }
+        else
+        {
+            /* Invalid header. Delete Frame. */
+            clearLocalRxBuffers();
+        }
+    }
+
+    /**
+     * Clear RX Buffer and counters.
+     */
+    void clearLocalRxBuffers()
+    {
+        memset(m_receiveFrame.raw, 0U, MAX_FRAME_LEN);
+        m_receivedBytes = 0U;
+        m_rxAttempts    = 0U;
+    }
+
+    /**
+     * Periodic heartbeat.
+     * Sends SYNC Command depending on the current Sync state.
+     * @param[in] currentTimestamp Time in milliseconds.
+     */
+    void heartbeat(const uint32_t currentTimestamp)
+    {
+        uint32_t heartbeatPeriod = HEATBEAT_PERIOD_UNSYNCED;
+
+        if (m_isSynced)
+        {
+            heartbeatPeriod = HEATBEAT_PERIOD_SYNCED;
+        }
+
+        if ((currentTimestamp - m_lastSyncCommand) >= heartbeatPeriod)
+        {
+            /* Timeout. */
+            if (m_lastSyncCommand != m_lastSyncResponse)
+            {
+                m_isSynced = false;
+            }
+
+            /* Send SYNC Command. */
+            uint8_t buf[CONTROL_CHANNEL_PAYLOAD_LENGTH] = {COMMANDS::SYNC};
+
+            /* Using (sizeof(buf) - 1U) as CMD Byte is not passed. */
+            Util::uint32ToByteArray(&buf[1], (sizeof(buf) - 1), currentTimestamp);

```

Comment:
the magic one :-)

Reply:
Is the first one meant, or the second one?